crates/red_knot_module_resolver/src/resolver.rs

```diff
@@ -1009,4 +1009,75 @@ mod tests {
 
         Ok(())
     }
+
+    #[test]
+    fn adding_or_removing_file_on_which_module_resolution_depends_invalidates_previously_successful_query_that_still_succeeds(
+    ) {
+        const TYPESHED: MockedTypeshed = MockedTypeshed {
+            versions: "functools: 3.8-",
+            stdlib_files: &[("functools.pyi", "def update_wrapper(): ...")],
+        };
+
+        let TestCase {
+            mut db,
+            src,
+            stdlib,
+            site_packages,
+            ..
+        } = TestCaseBuilder::new()
+            .with_custom_typeshed(TYPESHED)
+            .with_target_version(TargetVersion::Py38)
+            .build();
+
+        let functools_module_name = ModuleName::new_static("functools").unwrap();
+        let stdlib_functools_path = stdlib.join("functools.pyi");
+
+        let functools_module = resolve_module(&db, functools_module_name.clone()).unwrap();
+        assert_eq!(functools_module.search_path(), stdlib);
+        assert_eq!(
+            Some(functools_module.file()),
+            system_path_to_file(&db, &stdlib_functools_path)
+        );
+
+        // Adding a file to site-packages does not invalidate the query,
+        // since site-packages takes lower priority in the module resolution...
+        let site_packages_functools_path = site_packages.join("functools.py");
+        db.write_file(&site_packages_functools_path, "f: int")
+            .unwrap();
+        let functools_module = resolve_module(&db, functools_module_name.clone()).unwrap();
+        assert_eq!(functools_module.search_path(), stdlib);
+        assert_eq!(
+            Some(functools_module.file()),
+            system_path_to_file(&db, &stdlib_functools_path)
+        );
+
+        // ...but adding a first-party file does invalidate the query,
+        // since first-party files take higher priority in module resolution:
+        let src_functools_path = src.join("functools.py");
+        db.write_file(&src_functools_path, "FOO: int").unwrap();
+        let functools_module = resolve_module(&db, functools_module_name.clone()).unwrap();
+        assert_eq!(functools_module.search_path(), src);
+        assert_eq!(
+            Some(functools_module.file()),
+            system_path_to_file(&db, &src_functools_path)
+        );
+
+        // If we now delete both the stdlib and first-party files,
+        // it should resolve to site-packages:
+        for path in [stdlib_functools_path, src_functools_path] {
+            db.memory_file_system().remove_file(&path).unwrap();
+            File::touch_path(&mut db, &path);
+        }
+        let stdlib_versions_path = stdlib.join("VERSIONS");
+        db.memory_file_system()
+            .write_file(&stdlib_versions_path, "")
+            .unwrap();
+        File::touch_path(&mut db, &stdlib_versions_path);
```

Comment:
You can use `write_file` here
```suggestion
        db.write_file(&stdlib_versions_path, "")
            .unwrap();
```